mii/grpc_related/modelresponse_server.py

```diff
@@ -191,40 +199,82 @@ def intercept_service(self, continuation, handler_call_details):
         next_handler = continuation(handler_call_details)
         assert next_handler.unary_unary is not None
 
+        #USE KWARGS LIKE THEY ARE USED TO MAKE SESSIONS TO GET THE DEPLOYMENT NAME TO HASH THE COUNTERS/STUBS
+
         def invoke_intercept_method(request_proto, context):
             method_name = _get_grpc_method_name(handler_call_details.method)
-
             if method_name == TERMINATE_METHOD:
-                for stub in self.stubs:
-                    stub.invoke(TERMINATE_METHOD,
-                                google_dot_protobuf_dot_empty__pb2.Empty())
+                for deployment in self.stubs:
+                    for stub in self.stubs[deployment]:
+                        stub.invoke(TERMINATE_METHOD,
+                                    google_dot_protobuf_dot_empty__pb2.Empty())
                 self.asyncio_loop.call_soon_threadsafe(self.asyncio_loop.stop)
                 return next_handler.unary_unary(request_proto, context)
+            kwargs = unpack_proto_query_kwargs(request_proto.query_kwargs)
+            if method_name != TERMINATE_METHOD:
+                assert "deployment_name" in kwargs, "Must include deployment_name in kwargs for query"
+            deployment_name = kwargs.get('deployment_name')
+            kwargs.pop('deployment_name', None)
+            task = None
+            for repl in self.replica_configs:
+                if repl.deployment_name == deployment_name:
+                    task = repl.task
+                    break
```

Comment:
This should probably be in `__init__`. We store a dictionary mapping `deployment_name` to `task` and avoid doing this computation every query